docker-wrappers/SPRAS/README.md

```diff
@@ -0,0 +1,54 @@
+# SPRAS Docker image
+
+## Building
+A Docker image for SPRAS that is available on [Dockerhub]()
+This image comes bundled with all of the necessary software packages to run SPRAS, and can be used for execution in distributed environments (like HTCondor).
+
+To create the Docker image, make sure you are in this repository's root directory, and from your terminal run:
+```
+docker build -t reedcompbio/spras -f docker-wrappers/SPRAS/Dockerfile .
+```
+
+This will copy the entire SPRAS repository into the container and install SPRAS with `pip`. As such, any changes you've made to the current SPRAS repository will be reflected in version of SPRAS installed in the container. Since SPRAS
+is being installed with `pip`, it's also possible to specify that you want development modules installed as well. If you're using the container for development and you want the optional `pre-commit` and `pytest` packages as well as a
+spras package that receives changes without re-installation, change the
+`pip` installation line to:
+```
+pip install -e .[dev]
+```
+This will cause changes to spras source code to update the intsalled package.
+
+**Note:** This image will build for the same platform that is native to your system (ie amd64 or arm64). If you need to run this in a remote environment like HTCondor that is almost certainly `amd64` but you're building from Apple Silicon, it is recommended to either modify the Dockerfile to pin the platform:
+```
+FROM --platform=linux/amd64 almalinux:9
+```
+
+Or to temporarily override your system's default by exporting the environment variable:
+```
+export DOCKER_DEFAULT_PLATFORM=linux/amd64
+```
+(This environment variable can then be cleared by running `unset DOCKER_DEFAULT_PLATFORM` to return your system to its default)
+
+
+## Testing
```

Comment:
I submitted a test job following these instructions. They were clear to me, though I am experienced with HTCondor. The job is still idle.

Update: I saw too late that I need to change the `container_image` in the sub file for testing. I'll try again later.

Reply:
My updated job is still idle. This one uses `docker://reedcompbio/spras:latest`. `condor_q -better-analyze` reports

```
-- Schedd: ap2001.chtc.wisc.edu : <128.105.68.112:9618?...
The Requirements expression for job 130053.000 is

    (TARGET.PoolName == "CHTC") && ((HAS_SINGULARITY == true)) && TARGET.HasContainer && TARGET.HasDockerURL &&
    (TARGET.Disk >= RequestDisk) && (TARGET.Memory >= RequestMemory) && (TARGET.Cpus >= RequestCpus) && (TARGET.HasFileTransfer)

Job 130053.000 defines the following attributes:

    RequestCpus = 4
    RequestDisk = 16777216 (kb)
    RequestMemory = 4096 (mb)

The Requirements expression for job 130053.000 reduces to these conditions:

         Slots
Step    Matched  Condition
-----  --------  ---------
[1]           0  HAS_SINGULARITY == true

No successful match recorded.
Last failed match: Wed Mar 20 14:22:11 2024

Reason for last match failure: no match found

130053.000:  Run analysis summary ignoring user priority.  Of 447 machines,
    447 are rejected by your job's requirements
      0 reject your job because of their own requirements
      0 match and are already running your jobs
      0 match but are serving other users
      0 are able to run your job

WARNING:  Be advised:
   No machines matched the jobs's constraints
```

@jhiemstrawisc have you been able to run this in the CHTC pool?

Reply:
Interesting, I've only run on the PATh facility at this point (which was close enough to the true OSPool for my purposes, that I could tell). It's interesting that CHTC machines don't advertise singularity. I'll look into this.

Reply:
Sigh. It turns out the workflow can handle apptainer-in-apptainer, but apptainer doesn't work unless the launching user has an entry in `/etc/passwd`. CHTC launches this job under the Docker runtime with a random UID but doesn't the corresponding entry in `/etc/passwd`, so apptainer bricks. Without having to build an extra `.sif` for SPRAS itself, I'm trying to find a way to tell condor "yes, I'm giving you a docker URL, but I actually want you to launch the container with apptainer"